python/ray/serve/_private/http_proxy.py

```diff
@@ -650,6 +668,79 @@ async def proxy_asgi_receive(
             if msg["type"] == "websocket.disconnect":
                 return msg["code"]
 
+    async def _assign_request_with_timeout(
+        self,
+        handle: RayServeHandle,
+        scope: Scope,
+        disconnected_task: asyncio.Task,
+        timeout_s: Optional[float] = None,
+    ) -> Optional[StreamingObjectRefGenerator]:
+        """Attempt to send a request on the handle within the timeout.
+
+        If `timeout_s` is exceeded while trying to assign a replica, `TimeoutError`
+        will be raised.
+
+        `disconnected_task` is expected to be done if the client disconnects; in this
+        case, we will abort assigning a replica and return `None`.
+        """
+        assignment_task = handle.remote(pickle.dumps(scope), self.self_actor_handle)
+        done, _ = await asyncio.wait(
+            [assignment_task, disconnected_task],
+            return_when=FIRST_COMPLETED,
+            timeout=timeout_s,
+        )
+        if disconnected_task in done:
+            assignment_task.cancel()
+            return None
+        elif assignment_task in done:
+            return assignment_task.result()
+        else:
+            raise TimeoutError()
+
+    async def _consume_and_send_asgi_message_generator(
+        self,
+        obj_ref_generator: StreamingObjectRefGenerator,
+        send: Send,
+        timeout_s: Optional[float] = None,
+    ) -> Optional[str]:
+        """Consumes an obj ref generator that yields ASGI messages.
+
+        The messages are sent over the `send` interface.
+
+        If timeout_s is `None`, there's no timeout. If it's not `None`, a timeout error
+        will be raised if the full generator isn't consumed within the timeout.
+
+        Returns the status code for HTTP responses.
+        """
+        status_code = ""
+        start = time.time()
+        while True:
+            try:
+                if timeout_s is None:
+                    curr_timeout_s = None
+                else:
+                    curr_timeout_s = timeout_s - (time.time() - start)
+
+                obj_ref = await obj_ref_generator._next_async(timeout_s=curr_timeout_s)
```

Comment:
What's `obj_ref_generator`'s behavior if the timeout is negative? What if it's zero?

Reply:
Good question, let me check and ensure we're being defensive here.

Reply:
Couldn't `curr_timeout_s = timeout_s - (time.time() - start)` make `curr_timeout_s` negative?

Reply:
let me fix and add unit tests

Reply:
good catch. I added this to `utils.py::calculate_remaining_timeout` and added a test case for it.

Reply:
Nice, thanks!